src/libsyntax/parse/lexer/mod.rs

```diff
@@ -175,6 +175,16 @@ impl<'a> StringReader<'a> {
         self.fatal_errs.clear();
     }
 
+    pub fn buffer_fatal_errors(&mut self) -> Vec<Diagnostic> {
```

Comment:
Was this was the only API addition you needed?

Reply:
This and the `new_or_buffered_errs` constructor. I needed a way to "handle" the errors without prematurely emitting them (in this case, to cancel them), and this felt like a natural way to get the content of `self.fatal_errs` without just exposing the Vec directly. Since both `new` and `new_without_err` emitted the errors directly, i added this instead.

Reply:
I'm not against it, I'm surprised at how little you needed to add :)

Reply:
Yup! I just needed a version that wouldn't immediately crash the process if it ran into a problem in the first token. One thing i will ask is whether the current implementation with `buffer()` is appropriate, since plain Diagnostics don't have the destructor bomb that full `DiagnosticBuilder`s have. (Since they need a `Handler` to be emitted, they don't have the same ability to forcibly print themselves.) The alternative would be to swap out the Vec entirely and return the full DiagnosticBuilder, and since the `ParseSess` isn't local the lifetimes would work out.

Reply:
cc @petrochenkov